main_test.go

```diff
@@ -33,3 +40,194 @@ func TestBadArgs(t *testing.T) {
+	} else {
+		t.Fail()
+	}
+}
+
+func setupConfig(t *testing.T) (tempDir string) {
+	tempDir, err := ioutil.TempDir("", "config-test")
+	assert.NoError(t, err)
+	configuration.HomeDir = tempDir
+	return tempDir
+}
+
+func resetConfig(t *testing.T) {
+	var err error
+	configuration.HomeDir, err = os.UserHomeDir()
+	assert.NoError(t, err)
+}
+
+func TestDoConfigOssIndex(t *testing.T) {
+	tempDir := setupConfig(t)
+	defer resetConfig(t)
+	//noinspection GoUnhandledErrorResult
+	defer os.RemoveAll(tempDir) // clean up
+
+	var stdin bytes.Buffer
+	stdin.Write([]byte("ossindex\nmyOssiUsername\nmyOssiToken\n"))
+	err := doConfig(&stdin)
+	assert.NoError(t, err)
+}
+
+func TestDoConfigIq(t *testing.T) {
+	tempDir := setupConfig(t)
+	defer resetConfig(t)
+	//noinspection GoUnhandledErrorResult
+	defer os.RemoveAll(tempDir) // clean up
+
+	var stdin bytes.Buffer
+	stdin.Write([]byte("iq\nmyIqServer\nmyIqUsername\nmyIqToken\n"))
+	err := doConfig(&stdin)
+	assert.NoError(t, err)
+}
+
+func TestProcessConfigHelp(t *testing.T) {
+	err := processConfig(configuration.Configuration{Help: true})
+	assert.Error(t, err)
+	if exiterr, ok := err.(customerrors.ErrorExit); ok {
+		assert.Equal(t, 0, exiterr.ExitCode)
+	} else {
+		t.Fail()
+	}
+}
+
+func TestProcessConfigVersion(t *testing.T) {
+	err := processConfig(configuration.Configuration{Version: true})
+	assert.Error(t, err)
+	if exiterr, ok := err.(customerrors.ErrorExit); ok {
+		assert.Equal(t, 0, exiterr.ExitCode)
+	} else {
+		t.Fail()
+	}
+}
+
+func TestProcessConfigCleanCache(t *testing.T) {
+	err := processConfig(configuration.Configuration{CleanCache: true})
+	assert.NoError(t, err)
+}
+
+func TestCheckStdInInvalid(t *testing.T) {
+	err := checkStdIn()
+	if exiterr, ok := err.(customerrors.ErrorExit); ok {
+		assert.Equal(t, 1, exiterr.ExitCode)
+	} else {
+		t.Fail()
+	}
+}
+
+func TestCheckStdInValid(t *testing.T) {
+	tmpfile, err := ioutil.TempFile("", "fakeStdIn")
+	assert.Nil(t, err)
+	//noinspection GoUnhandledErrorResult
+	defer os.Remove(tmpfile.Name()) // clean up
+
+	content := []byte("yadda\n")
+	if _, err := tmpfile.Write(content); err != nil {
+		assert.NoError(t, err)
+	}
+	if _, err := tmpfile.Seek(0, 0); err != nil {
+		assert.NoError(t, err)
+	}
+	oldStdin := os.Stdin
+	defer func() { os.Stdin = oldStdin }() // Restore original Stdin
+	os.Stdin = tmpfile
+
+	err = checkStdIn()
+	assert.NoError(t, err)
+}
+
+func TestCheckOSSIndexNoneVulnerable(t *testing.T) {
+	// TODO find way to mock ossindex url for this test, probably just move checkOSSIndex() method to ossindex package
+	purls := []string{"pkg:github/BurntSushi/toml@0.3.1"}
+	invalidPurls := []string{"invalidPurl"}
+	noColor := true
+	quiet := true
+	config := configuration.Configuration{Formatter: &audit.AuditLogTextFormatter{Quiet: &quiet, NoColor: &noColor}}
+	err := checkOSSIndex(purls, invalidPurls, config)
+	assert.NoError(t, err)
+}
+
+func TestDoOssi(t *testing.T) {
```

Comment:
I am always SUPER wary of testing private methods. Do you think we really need to do this, versus an Integration test via dogfooding, etc... that tests more "public" expected behavior?

Reply:
> I am always SUPER wary of testing private methods. Do you think we really need to do this, versus an Integration test via dogfooding, etc... that tests more "public" expected behavior?

I felt naked trying to change all the `os.Exit` calls with neither public IT's, nor unit tests of the code being changed. I therefore added some very specific unit tests to enable the refactoring at hand. I agree that IT's at a higher level would be nice to have, but I also believe this is a "chicken and the egg" problem - can't write the high level IT's with all those pesky `os.Exit` calls. Can't do true refactoring (where "refactoring" means making changes that can be proven not to break existing behavior) without unit tests or IT's.

So, I'd be happy to see unit tests be replaced by higher level IT's, which should hopefully be easier to do once the `os.Exit` beast is slain.

Reply:
I'll see if I can noodle out a proper IT framework. I agree it would be super useful.